### ISSUE TITLE
BDD configuration comments should be in XML.

### ISSUE THREAD

**Issue by nrmancuso** [checkstyle/checkstyle#12155]
Consider:

```bash
~/I/checkstyle/s/t/resources/c/p/t/c/checks/im/importorder master *1 !2 > cat InputImportOrderStaticGroupOrderBottom_Negative2.java
/*
ImportOrder
option = top
groups = org, java
ordered = (default)true
separated = true
separatedStaticGroups = (default)false
caseSensitive = (default)true
staticGroups = (default)
sortStaticImportsAlphabetically = true
useContainerOrderingForStatic = (default)false
tokens = (default)STATIC_IMPORT


*/

package com.puppycrawl.tools.checkstyle.checks.imports.importorder;

import org.antlr.v4.runtime.*; // ok

import static java.lang.Math.PI; // violation
import static org.antlr.v4.runtime.Recognizer.EOF; // ok

import java.util.Set; // ok

public class InputImportOrderStaticGroupOrderBottom_Negative2
{

}

```

If I want to test or use this config locally or set up an example for an issue report, I need to manually translate this into a configuration module:

```
~/IdeaProjects/checkstyle_tester/src > cat import_config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="ImportOrder">
          <property name="option" value="top"/>
          <property name="groups" value="org, java"/>
          <property name="ordered" value="true"/>
          <property name="separated" value="true"/>
          <property name="separatedStaticGroups" value="false"/>
          <property name="caseSensitive" value="true"/>
          <property name="staticGroups" value=""/>
          <property name="sortStaticImportsAlphabetically" value="true"/>
          <property name="useContainerOrderingForStatic" value="false"/>
          <property name="tokens" value="STATIC_IMPORT"/>
        </module>
    </module>
</module>


```

I am proposing that we migrate from our custom config/ config parser, to simply extract xml from block comment in header to create check module configs.

Example:
```
~/I/checkstyle/s/t/resources/c/p/t/c/checks/im/importorder master *1 !2 > cat InputImportOrderStaticGroupOrderBottom_Negative2.java
/*
<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="ImportOrder">
          <property name="option" value="top"/>
          <property name="groups" value="org, java"/>
          <property name="ordered" value="true"/>
          <property name="separated" value="true"/>
          <property name="separatedStaticGroups" value="false"/>
          <property name="caseSensitive" value="true"/>
          <property name="staticGroups" value=""/>
          <property name="sortStaticImportsAlphabetically" value="true"/>
          <property name="useContainerOrderingForStatic" value="false"/>
          <property name="tokens" value="STATIC_IMPORT"/>
        </module>
    </module>
</module>
*/

package com.puppycrawl.tools.checkstyle.checks.imports.importorder;

import org.antlr.v4.runtime.*; // ok

import static java.lang.Math.PI; // violation
import static org.antlr.v4.runtime.Recognizer.EOF; // ok

import java.util.Set; // ok

public class InputImportOrderStaticGroupOrderBottom_Negative2
{

}

```

Using xml for our configuration header has several advantages:
1. Allows us to validate syntax/ format of configuration easily
2. Will ease issues where we need multiple modules, etc.
3. Can even allow comments inline to explain complicated module combinations in place
4. Provides portability; the same config used in a given test can be used in an actual configuration file, issue report, etc.
5. Allows us to remove all current limitations of BDD.

**Comment by romani** [comment]
>Allows us to validate syntax/ format of configuration easily

We already payed a price for this, it is not an issue now.

>Will ease issues where we need multiple modules

We do not want to have complicated configs, issue with filters (where we need few modules) looks like resolved.

>Can even allow comments inline to explain complicated module combinations in place

I might be wrong, but , content of comment is property file, so it should support `#`

>Allows us to remove all current limitations of BDD.

Probably good time to list them all.

>same config used in a given test can be used in an actual configuration file

Please share more on this, in what cases you need this.

----------

We had idea of xml at the beginning, but abandoned it due to high technical verbosity. 
Issues with length of lines.
Reviewer should look at config content easily.

**Comment by romani** [comment]
We can try this as extra option.
We can have comment in 
```
/*xml

*/
```

To let parser do load by xml.
So there will be no conflicts, and it might find its place in life.

**Comment by stoyanK7** [comment]
> We can try this as extra option.

The idea looks good at first sight. I think if we go with this, the parser for the examples will be the testing ground before migrating all tests? Or is the idea to make and use this only for the examples parser?

**Comment by nrmancuso** [comment]
Hopefully it is testing ground :)

**Comment by romani** [comment]
We will not migrate all tests.
All we can do is to use it in ExampleX.java.

**Comment by stoyanK7** [comment]
@romani @nrmancuso @rnveach Do you have any tips on how this is going to work?

I am digging right now. I  imagine a new method `verifyWithXmlConfigParser()` that passes file to `XmlConfigParser`(new class in `bdd` package) which will extract config from comment starting with `/*xml` and finishing with `*/`. Then `verifyWithXmlConfigParser()` passes that configuration to `ConfigurationLoader.loadConfiguration()`.

This is what I imagine after just looking around. I could be completely wrong.

**Comment by romani** [comment]
I thought in this as update of existing verify method with internal switch/if to load content as xml instead of as properties. 
We just need to have ready to use config object to execute. How this config was loaded is business on very specific phase in this verify method.

In worse case I ok with verifyWithXmlConfigParser()

**Comment by nrmancuso** [comment]
Test methods should not have to care about which verify method to use.  Either try and fail to parse one type and the try other, or use special comment format like ‘/*xml’

**Comment by romani** [comment]
Better to use some marker to be explicit
